File: Backtracking/Combinations2/comb.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void doRecursion(vector<vector<int>>& ans,
                     vector<int> helper,
                     int n,
                     int k) {

        if(helper.size() == k) {
            int check = 0;

            for(int i = 0; i < k; i++) {
                check += helper[i];
            }

            if(check == n) {
                ans.push_back(helper);
            }

            return;
        }

        int size = helper.size();

        for(int i = 1; i <= 9; i++) {

            if(size > 0 && helper[size - 1] < i) {
                helper.push_back(i);
                doRecursion(ans, helper, n, k);
                helper.pop_back();
            }

            if(size == 0) {
                helper.push_back(i);
                doRecursion(ans, helper, n, k);
                helper.pop_back();
            }
        }
    }

    vector<vector<int>> combinationSum(int n, int k) {
        vector<vector<int>> ans;

        doRecursion(ans, {}, n, k);

        return ans;
    }
};

File: Backtracking/Combinations2/comb.py
class Solution:
    
    def doRecursion(self, ans, helper, n, k):
        if len(helper) == k:
            check = 0
            for i in range(k):
                check += helper[i]
            if check == n:
                ans.append(helper.copy())
            return
    
        size = len(helper)
    
        for i in range(1, 10):
            if size > 0 and helper[size - 1] < i:
                helper.append(i)
                self.doRecursion(ans, helper, n, k)
                helper.pop()
    
            if size == 0:
                helper.append(i)
                self.doRecursion(ans, helper, n, k)
                helper.pop()

    def combinationSum(self, n, k):
        # code here
        ans = []
        self.doRecursion(ans, [], n, k)
        return ans